### PR TITLE
Layout-generalized concat_on_depth

### PR DESCRIPTION
**Pull request by primenumber**
Support mixed layout (e.g. ChHWBCl and HWChBCl) concat_on_depth.

## Motivation and Context

There are several memory layout for Quantized and packed types.
And these layouts may be mixed in a network, even in inputs for an ops.
That cause compile error in concat_on_depth.
This PR fix it.
If you want to add new layout, you have to just implement index_{height,width,channels_high} and (access_ChHWBCl or access_NHWC).

## Description

- Re-implement func_ConcatOnDepth

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: dlk/python/dlk/templates/include/func/concat_on_depth.h [hunk 1]
```diff
@@ -16,90 +16,153 @@ limitations under the License.
...
+    case MemoryLayout::ChHWBCl: return 2;
+    case MemoryLayout::HWChBCl: return 1;
+    case MemoryLayout::NHWC: return 2;
+    default: return 0;
+  }
+}
 
-  Measurement::Stop();
+template<class T>
+T& access_ChHWBCl(const TensorView<T, MemoryLayout::ChHWBCl>& tensor,
```
Comment (tkng) [inline dlk/python/dlk/templates/include/func/concat_on_depth.h]: This function looks existing to access quantized tensor element, but the actual memory layout depends on the type  of the argument `tensor`. How about changing the function name to `access_quantized_tensor_elem` or something?

Just `access` or `access_tensor_element` might be enough.

File: dlk/python/dlk/templates/include/func/concat_on_depth.h [hunk 2]
```diff
@@ -16,90 +16,153 @@ limitations under the License.
...
+    const std::size_t digit) {
+  return tensor(h, w, ch, digit, 0);
+}
 
-  T_UINT depths[32];
-  for (T_UINT n = 0; n < n_inputs; ++n) {
-    depths[n] = inputs[n].get_shape()[0];
-  }
+template<class T>
+T& access_NHWC(const TensorView<T, MemoryLayout::NHWC>& tensor,
```
Comment (tkng) [inline dlk/python/dlk/templates/include/func/concat_on_depth.h]: I propose to change the function name for the same reason as above. Just `access `or `access_tensor_element might` be enough.

**Comment by tkng** [comment]
The new code looks complex, it took a while for me to understand that the new code is calling the template class recursively to concat arbitrary number of tensors. However, I have no good counter proposal.

**Comment by primenumber** [comment]
I agree that new code is complex.
If we introduce C++17, we can use `if constexpr` or `fold expression` to reduce complexity...

File: dlk/python/dlk/templates/include/func/concat_on_depth.h (hunk 1 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/func/concat_on_depth.h]: OK, fixed.

File: dlk/python/dlk/templates/include/func/concat_on_depth.h (hunk 2 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/func/concat_on_depth.h]: also fixed.

**Comment by primenumber** [comment]
run lmnet test

**Comment by tkng** [comment]
Now I'm fixing test fail problem, that will be merged soon. After that, you can re-run the test.

> If we introduce C++17, we can use if constexpr or fold expression to reduce complexity...

Considering that even out docker image is based on Ubuntu 16.04, we cannot use C++ 17 for now.
Maybe it's a good idea to leave a comment like "TODO: after compiler update to C++17, rewrite this function with if constexpr of fold expression."

**Pull request merged**